File: packages/python/src/telebot_py/bot/bulk.py
"""Bulk message Bot API methods (parity with packages/go/pkg/bot/bulk.go)."""

from __future__ import annotations

from collections.abc import Sequence

from telebot_py.bot.base import Requester, clean_payload, parse_flag, parse_list_result
from telebot_py.types.common import MessageId


class BulkMixin(Requester):
    """Bot methods for forwarding, copying, and deleting many messages at once."""

    async def forward_messages(
        self,
        chat_id: int | str,
        from_chat_id: int | str,
        message_ids: Sequence[int],
        *,
        business_connection_id: str | None = None,
        message_thread_id: int | None = None,
        disable_notification: bool | None = None,
        protect_content: bool | None = None,
    ) -> list[MessageId]:
        """Forward multiple messages of any kind to a target chat.

        Example:
            >>> ids = await bot.forward_messages(123, "@source", [11, 12])

        Args:
            chat_id: Unique identifier for the target chat or username of the
                target channel.
            from_chat_id: Unique identifier of the source chat or username of
                the source channel.
            message_ids: Identifiers of the messages to forward, 1-100.
            business_connection_id: Business connection on whose behalf to act.
            message_thread_id: Target message thread identifier.
            disable_notification: Send silently.
            protect_content: Protect the forwarded content from saving.

        Returns:
            The MessageId list of the forwarded messages.

        Raises:
            InvalidTokenError: If Telegram rejects the token (HTTP 401).
            TelegramApiError: If Telegram responds not-ok or retries exhaust.
            NetworkError: If the transport keeps failing after retries.
        """
        payload = clean_payload(
            chat_id=chat_id,
            from_chat_id=from_chat_id,
            message_ids=list(message_ids),
            business_connection_id=business_connection_id,
            message_thread_id=message_thread_id,
            disable_notification=disable_notification,
            protect_content=protect_content,
        )
        return parse_list_result(MessageId, await self.request("forwardMessages", payload))

    async def copy_messages(
        self,
        chat_id: int | str,
        from_chat_id: int | str,
        message_ids: Sequence[int],
        *,
        business_connection_id: str | None = None,
        message_thread_id: int | None = None,
        disable_notification: bool | None = None,
        protect_content: bool | None = None,
        remove_caption: bool | None = None,
    ) -> list[MessageId]:
        """Copy multiple messages without linking to the originals.

        Example:
            >>> ids = await bot.copy_messages("@target", 456, [1, 2])

        Args:
            chat_id: Unique identifier for the target chat or username of the
                target channel.
            from_chat_id: Unique identifier of the source chat or username of
                the source channel.
            message_ids: Identifiers of the messages to copy, 1-100.
            business_connection_id: Business connection on whose behalf to act.
            message_thread_id: Target message thread identifier.
            disable_notification: Send silently.
            protect_content: Protect the copied content from saving.
            remove_caption: Pass True to copy messages without their captions.

        Returns:
            The MessageId list of the copied messages.

        Raises:
            InvalidTokenError: If Telegram rejects the token (HTTP 401).
            TelegramApiError: If Telegram responds not-ok or retries exhaust.
            NetworkError: If the transport keeps failing after retries.
        """
        payload = clean_payload(
            chat_id=chat_id,
            from_chat_id=from_chat_id,
            message_ids=list(message_ids),
            business_connection_id=business_connection_id,
            message_thread_id=message_thread_id,
            disable_notification=disable_notification,
            protect_content=protect_content,
            remove_caption=remove_caption,
        )
        return parse_list_result(MessageId, await self.request("copyMessages", payload))

    async def delete_messages(
        self,
        chat_id: int | str,
        message_ids: Sequence[int],
        *,
        business_connection_id: str | None = None,
    ) -> bool:
        """Delete multiple messages from a chat at once.

        Example:
            >>> ok = await bot.delete_messages(789, [5, 6])

        Args:
            chat_id: Unique identifier for the target chat or username of the
                target channel.
            message_ids: Identifiers of the messages to delete, 1-100.
            business_connection_id: Business connection on whose behalf to act.

        Returns:
            True on success.

        Raises:
            InvalidTokenError: If Telegram rejects the token (HTTP 401).
            TelegramApiError: If Telegram responds not-ok or retries exhaust.
            NetworkError: If the transport keeps failing after retries.
        """
        payload = clean_payload(
            chat_id=chat_id,
            message_ids=list(message_ids),
            business_connection_id=business_connection_id,
        )
        return parse_flag(await self.request("deleteMessages", payload))

File: packages/go/pkg/bot/bulk.go
package bot

import (
	"context"

	"github.com/Nam088/telebot/packages/go/pkg/types"
)

// ForwardMessages forwards multiple messages of any kind to a target chat.
//
// Parameters:
//   - ctx: Context for request cancellation and timeout.
//   - opts: Forward options including chat_id, from_chat_id, and message_ids.
//
// Returns:
//   - []types.MessageId: Identifiers of the forwarded messages on success.
//   - error: TelegramError if the API returns an error, or a network error.
//
// Telegram API: https://core.telegram.org/bots/api#forwardmessages
func (b *Bot) ForwardMessages(ctx context.Context, opts *types.ForwardMessagesOptions) ([]types.MessageId, error) {
	var ids []types.MessageId
	if err := b.Request(ctx, "forwardMessages", opts, &ids); err != nil {
		return nil, err
	}
	return ids, nil
}

// CopyMessages copies multiple messages of any kind without linking to the originals.
//
// Parameters:
//   - ctx: Context for request cancellation and timeout.
//   - opts: Copy options including chat_id, from_chat_id, and message_ids.
//
// Returns:
//   - []types.MessageId: Identifiers of the copied messages on success.
//   - error: TelegramError if the API returns an error, or a network error.
//
// Telegram API: https://core.telegram.org/bots/api#copymessages
func (b *Bot) CopyMessages(ctx context.Context, opts *types.CopyMessagesOptions) ([]types.MessageId, error) {
	var ids []types.MessageId
	if err := b.Request(ctx, "copyMessages", opts, &ids); err != nil {
		return nil, err
	}
	return ids, nil
}

// DeleteMessages deletes multiple messages from a chat at once.
//
// Parameters:
//   - ctx: Context for request cancellation and timeout.
//   - opts: Delete options including chat_id and message_ids.
//
// Returns:
//   - bool: True on success.
//   - error: TelegramError if the API returns an error, or a network error.
//
// Telegram API: https://core.telegram.org/bots/api#deletemessages
func (b *Bot) DeleteMessages(ctx context.Context, opts *types.DeleteMessagesOptions) (bool, error) {
	var ok bool
	if err := b.Request(ctx, "deleteMessages", opts, &ok); err != nil {
		return false, err
	}
	return ok, nil
}
